drivers/net/ppp.c

```diff
@@ -416,8 +416,30 @@ static void ppp_process_msg(struct ppp_driver_context *ppp)
 #endif
 			net_pkt_unref(ppp->pkt);
 		} else {
-			/* Skip FCS bytes (2) */
-			net_buf_frag_last(ppp->pkt->buffer)->len -= 2;
+			/* Remove FCS bytes from the end (2) */
+			struct net_buf *parent = NULL;
+			struct net_buf *cur = ppp->pkt->buffer;
+
+			while (cur->frags) {
+				parent = cur;
+				cur = cur->frags;
+			}
+			if (cur->len >= 2) {
+				cur->len -= 2;
+			} else {
+				if (cur->len == 1) {
+					cur->len--;
+					if (parent && parent->len > 1) {
+						parent->len--;
+					}
+				} else {
+					if (parent && parent->len > 2) {
+						parent->len -= 2;
+					}
+				}
+				if (parent)
+					net_pkt_frag_del(ppp->pkt, parent, cur);
+			}
```

Comment:
This seems too complicated to me. First of all, make sure that this implementation details are in a separate "helper" function. Second, what about making it a bit more generic:

```
/*
 * Truncate 'len' bytes from tail of net_pkt, assuming that requested remove
 * area spans maximum of 2 fragments.
 */
static void net_pkt_truncate_tail(struct net_pkt *pkt, size_t len)
{
	struct net_buf *buf = pkt->frags;
	struct net_buf *prev = NULL;

	/* find two last fragments */
	while (buf->frags) {
		prev = buf;
		buf = buf->frags;
	}

	/* truncate last packet */
	if (buf->len >= len) {
		buf->len -= len;
		return;
	}

	/* remove last packet */
	net_buf_frag_del(prev, buf);

	/* truncate "new" last packet */
	prev->len -= len;
}
```

Reply:
Agreed on making a helper.
That suggested code appears to have at least two bugs though:
- If `buf->len < len`, you'll remove the full `len` from `prev`, where you should have removed `len - buf->len`
- If `prev` has less than `len` bytes, you'll underflow
I'm not sure that the helper should be limited to a max of 2 fragments, either.  We can't walk the fragment list backwards, but you could just loop back to the top and scan the list again -- it's slow but would only happen in the case that would otherwise not work at all.

Reply:
>If buf->len < len, you'll remove the full len from prev, where you should have removed len - buf->len

Yeah, I didn't test it. You are right.

>If prev has less than len bytes, you'll underflow

I don't think this is possible in this file context. I was assuming that this will be a static function used only in PPP context.

>I'm not sure that the helper should be limited to a max of 2 fragments, either. We can't walk the fragment list backwards, but you could just loop back to the top and scan the list again -- it's slow but would only happen in the case that would otherwise not work at all.

I was thinking about that. The function that I proposed is a compromise between begin generic, simple and most importatnly solve the task of removing small portion (less than a net_buf can contain) od data from net_buf fragments.

We can implement a generic function that will work with any set of parameters and any type of fragment list (also such that contains only one byte in a single net_buf fragment). My idea was to reverse net_buf fragment list by manipulating `net_buf->frags`, then calling something like `net_pkt_pull` API and then reverse net_buf fragment list once again. This would allow to keep `O(n)`.

Reply:
I think we can do it more easily in `O(n)`: walk fragments and count off `net_pkt_get_len(pkt) - len` bytes, then truncate wherever we run out, and delete fragments after that.

Reply:
@jimparis, @mniestroj: What's the status of this? @mniestroj, if you think that factoring out that code as a separate static function is really required, please give -1 until it's done. Otherwise, I'm keen to give my +1.

But also a friendly reminder that without tests for such cases, we'll be drowned in issues like this (newer changes regress/mis-interact with previous changes, etc.). (Btw, that would be a good "pro" for having a generic/non-static function like that, then we can unit-test it separately, which should be easier than "integrated" test with ppp-specific handling.)

Reply:
I think that this implementation is too complicated, but I won't block this PR (by -1) because of it. It is tricky to follow so many if-statements (there are simply too many branches) and there is no test for it.

I like @jimparis idea of:

>I think we can do it more easily in O(n): walk fragments and count off net_pkt_get_len(pkt) - len bytes, then truncate wherever we run out, and delete fragments after that.

together with @pfalcon suggestion of:

>Btw, that would be a good "pro" for having a generic/non-static function like that, then we can unit-test it separately, which should be easier than "integrated" test with ppp-specific handling.

which should give something like:

```
int net_pkt_remove_tail(struct net_pkt *pkt, size_t length)
{
	struct net_buf *buf = pkt->buffer;
	size_t pkt_len = net_pkt_get_len(pkt);

	if (pkt_len < length) {
		return -EINVAL;
	}

	pkt_len -= length;

	while (buf) {
		if (buf->len >= pkt_len) {
			buf->len = pkt_len;

			if (buf->frags) {
				net_pkt_frag_unref(buf->frags);
				buf->frags = NULL;
			}

			break;
		}

		buf = buf->frags;
		pkt_len -= buf->len;
	}

	return 0;
}
```

and easy way to add unit tests for it.

Reply:
> I think that this implementation is too complicated, but I won't block this PR (by -1) because of it. 

Right, it's a fix for obvious bug after all. Ping @jimparis, can you please do suggested refactorings, to make sure this fits into 2.6 (the freeze is soon AFAIK).

Reply:
> 		if (buf->len >= pkt_len) {
> 			buf->len = pkt_len;

Hmm, this does not look right. The `buf->len` is just the length of this small fragment, and we are comparing it to the total length of the pkt. I think we should compare to cumulative length of the pkt so far.
Also, setting `buf->len = pkt_len;` would set the length of this fragment to the total length of the pkt which would be wrong if there are multiple fragments in this chain.

Reply:
@jukkar Perhaps `pkt_len` is a bad variable name then. Maybe `remaining_len` would better match implemented logic. Then:
```
if (buf->len >= remaining_len) {
  	buf->len = remaining_len;
```
Notice that after entering `if (buf->len >= remaining_len) {` we already are at the last buffer that should remain, all the other buffers should be dropped. So `buf->len = remaining_len;` is either reducing length of last buffer or leaving it as is.

>Also, setting buf->len = pkt_len; would set the length of this fragment to the total length of the pkt which would be wrong if there are multiple fragments in this chain.

That is why `remaining_len` would be probably better name, as we decrease that variable in a loop by length of buffers in packet. So that variable actually represents how many bytes should still be left in packet.

Reply:
> Maybe `remaining_len` would better match implemented logic

Ack to that :)